Fix AI queueing structure build orders twice at higher game speeds due to lag

reaperrr opened the pull request:
Fixes #9403.

reaperrr committed:
Fix AI queueing build orders twice at higher game speeds due to lag

Fixes #9403.

penev92 commented:
This is a straight nerf for the AI :(
Can we just remove this property and replace
`waitTicks = active ? ai.Info.StructureProductionActiveDelay + randomFactor`
with
`waitTicks = active ? 3 * world.OrderLatency + randomFactor` ?

reaperrr commented:
> waitTicks = active ? 3 * world.OrderLatency + ai.Info.StructureProductionActiveDelay + randomFactor

Done. Although I set the default to 2 instead of 0 because it (very rarely) still happens at fastest speed.

penev92 commented:
You should revert/update the Desc on the property now. And maybe put it as a comment on the `3 * ...` to explain what that magic number is.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -56,13 +56,14 @@ public sealed class HackyAIInfo : IBotInfo, ITraitInfo
-		public readonly int StructureProductionActiveDelay = 10;
+		[Desc("Additional delay (in ticks) added between structure production checks when actively building things.",
+			"Note: The total delay is gamespeed OrderLatency x 3 + this + StructureProductionRandomBonusDelay.",
+			"A value of at least 2 or 3 is recommended to avoid AI issuing duplicate build orders.")]
```
Is this last sentence still true? That should be completely solved by the OrderLatency change.

pchote commented:
:+1: after addressing my question above.

reaperrr commented:
@pchote:
> Is this last sentence still true? That should be completely solved by the OrderLatency change.

Unfortunately it was still true, I still encountered it at fastest gamespeed with OrderLatency * 3. Only very rarely, but not never.

However, after some pondering I've now removed that part of the description, changed the active delay default to 0 instead of 2, and instead increased the hardcoded minimum to 4*OrderLatency.

Compared to before this update, this only increases the total delay by 1 tick at slow and normal speed (4x3 instead of 3x3 +2), 2 ticks at 'faster' (4x4 instead of 3x4 +2) and 4 at 'fastest' where the issue showed up most (6x4 instead of 6x3 +2). Compared to bleed it's +2 / +2 / +6 / +14, respectively.

That way the effect on AI efficiency should be negligible at slow and normal and offset by human player's limits at faster and fastest, and the duplication issue should be gone for good*.

*fingers crossed.

abcdefg30 commented:
Seems to fix the issue. :+1: / :white_check_mark:

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1a216bd80f339fca04b3ff291b23604b8abca872...28c9216915c3e86cbeff49b246524725d491e2e9)
